Support for cypher-dsl 2023.0.0

We're having an application where we both depend on neo4j-graphql-java and cypher-dsl.

Today, we're not able to update to cypher-dsl 2023.0.0 because of the following errors when the query is generated : 

```
at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.NoSuchMethodError: 'org.neo4j.cypherdsl.core.StatementBuilder$OngoingReadingWithWhere org.neo4j.cypherdsl.core.StatementBuilder$OngoingReadingWithoutWhere.where(org.neo4j.cypherdsl.core.Condition)'
	at org.neo4j.graphql.handler.QueryHandler.generateCypher(QueryHandler.kt:121)
	at org.neo4j.graphql.handler.BaseDataFetcher.get(BaseDataFetcher.kt:27)
	at org.neo4j.graphql.handler.BaseDataFetcher.get(BaseDataFetcher.kt:18)
```

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -18,7 +18,7 @@
         <dependency>
             <groupId>org.neo4j.driver</groupId>
             <artifactId>neo4j-java-driver</artifactId>
-            <version>5.2.0</version>
+            <version>5.3.1</version>
             <scope>test</scope>
         </dependency>
         <dependency>
@@ -73,7 +73,7 @@
         <dependency>
             <groupId>org.neo4j</groupId>
             <artifactId>neo4j-cypher-dsl</artifactId>
-            <version>2022.8.1</version>
+            <version>2023.0.0</version>
         </dependency>
         <dependency>
             <groupId>ch.qos.logback</groupId>

--- a/core/src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt
+++ b/core/src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt
@@ -74,7 +74,7 @@ class OptimizedFilterHandler(val type: GraphQLFieldsContainer, schemaConfig: Sch
             private val matchQueryWithoutWhere: OngoingReading,
             private val type: GraphQLFieldsContainer,
             private val value: Map<*, *>,
-            private val parentPassThroughWiths: Collection<Expression>,
+            private val parentPassThroughWiths: Collection<IdentifiableElement>,
             private val variables: Map<String, Any>
     ) {
 
@@ -156,7 +156,7 @@ class OptimizedFilterHandler(val type: GraphQLFieldsContainer, schemaConfig: Sch
         private fun handleQuantifierPredicates(
                 query: OrderableOngoingReadingAndWithWithoutWhere,
                 relFilter: RelationPredicate,
-                levelPassThroughWiths: LinkedHashSet<Expression>
+                levelPassThroughWiths: LinkedHashSet<IdentifiableElement>
         ): OrderableOngoingReadingAndWithWithoutWhere {
             val objectField = relFilter.value
             val nestedParsedQuery = parseFilter(objectField as Map<*, *>, relFilter.fieldDefinition.type.getInnerFieldsContainer())
@@ -227,7 +227,7 @@ class OptimizedFilterHandler(val type: GraphQLFieldsContainer, schemaConfig: Sch
         private fun createAdditionalConditions(
                 query: ParsedQuery,
                 relVariable: Node,
-                passThroughWiths: LinkedHashSet<Expression>,
+                passThroughWiths: LinkedHashSet<IdentifiableElement>,
                 filter: List<Pair<SymbolicName, AliasedExpression>>,
                 whereClauseFactory: WhereClauseFactory
         ): ConditionBuilder {
@@ -249,7 +249,7 @@ class OptimizedFilterHandler(val type: GraphQLFieldsContainer, schemaConfig: Sch
 
         private fun withClauseWithOptionalDistinct(
                 exposesWith: ExposesWith,
-                withs: Collection<Expression>,
+                withs: Collection<IdentifiableElement>,
                 useDistinct: Boolean = true) = when {
             useDistinct && withs.size == 1 -> exposesWith.withDistinct(*withs.toTypedArray())
             else -> exposesWith.with(*withs.toTypedArray())

--- a/core/src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt
+++ b/core/src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt
@@ -401,7 +401,7 @@ open class ProjectionBase(
     }
 
     fun cypherDirective(ctxVariable: SymbolicName, fieldDefinition: GraphQLFieldDefinition, arguments: Map<String, Any>, cypherDirective: CypherDirective, thisValue: SymbolicName?): ResultStatement {
-        val args = sortedMapOf<String, Expression>()
+        val args = sortedMapOf<String, IdentifiableElement>()
         if (thisValue != null) args["this"] = thisValue.`as`("this")
         arguments
             .filterNot { (name, _) -> SPECIAL_FIELDS.contains(name) }

